R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
 
   out <- nq / n * out
 }
+
+#' @title (Internal) Design-based variance estimates with HC0 adjustment
+#' @param x a fitted \code{teeMod} model
+#' @details The design-based variance estimates can be calculated for 
+#' \code{teeMod} models satisfying the following requirements:
+#' - The model uses \code{rct_design} as \code{Design}
+#' - The model only estimates a main treatment effect
+#' - Inverse probability weighting is incorporated
+#' 
+#' @keywords internal
+#' @rdname var_estimators
+.vcov_DB0 <- function(x, ...) {
+  if (!inherits(x, "teeMod")) {
+    stop("x must be a teeMod model")
+  }
+  
+  if (!x@lmitt_fitted){
+    # x@lmitt_fitted is false if someone created x using as.lmitt
+    stop("x must have been fitted using lmitt.formula")
+  }
+  
+  if (inherits(os <- x$model$`(offset)`, "SandwichLayer"))
+    if (!all(os@keys$in_Q)){
+      stop(paste("Design-based standard errors cannot be computed for teeMod",
+                 "models with external sample for covariance adjustment"))
+    }
+  
+  if (x@Design@type != "RCT"){
+    stop("x must be an RCT design")
+  }
+  
+  if (length(x@moderator) > 0){
+    stop(paste("Design-based standard errors cannot be computed for teeMod",
+               "models with moderators"))
+  }
+  
+  args <- list(...)
+  if ("type" %in% names(args)) {
+    stop(paste("Cannot override the `type` argument for meat",
+               "matrix computations"))
+  }
+  args$x <- x
+  args$db <- TRUE
+  n <- length(args$cluster)
+  
+  if (x@absorbed_intercepts) {
+    a22inv <- sandwich::bread(x)
+    meat <- do.call(sandwich::meatCL, args)
+    
+    vmat <- (1 / n) * a22inv %*% meat %*% a22inv
+    name <- grep("z", colnames(x$model))
```

Comment:
I don't think the treatment column will always be named `z`. I would find the exact treatment name using `var_names(x@Design, "t")`.

Reply:
This code attempts to align the row and column names of the design-based output of `vcov_tee` to model-based outputs. It seems that `z.` is currently used (for the main effect), which is what I'm trying to get to for the `name` variable. Is there a better way to do that?

Reply:
Use `var_names(x@Design, "t")`. `lmitt()` adds a `.` to the end of the treatment column found in the Design:
```r
data(simdata)
simdata$another_trt <- simdata$z
des <- rct_design(another_trt ~ unitid(uoa1, uoa2), simdata)
 lmitt(y ~ 1, design = des, data = simdata)
another_trt. 
      0.1205 
```

Reply:
Thanks! Updated in commit https://github.com/benbhansen-stats/propertee/pull/185/commits/7fbf8e2e103f0d7191d96ceb45987d08318f1aed